test/test-tcp-close-reset.c

```diff
@@ -0,0 +1,293 @@
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include "uv.h"
+#include "task.h"
+
+#include <errno.h>
+#include <string.h> /* memset */
+
+#define NUM_WRITE_REQS 4
+
+static uv_loop_t* loop;
+static uv_tcp_t tcp_server;
+static uv_tcp_t tcp_client;
+static uv_tcp_t tcp_accepted;
+static uv_connect_t connect_req;
+static uv_shutdown_t shutdown_req;
+
+static int client_close;
+static int shutdown_before_close;
+
+static int write_cb_called;
+static int close_cb_called;
+static int shutdown_cb_called;
+
+static void connect_cb(uv_connect_t* req, int status);
+static void write_cb(uv_write_t* req, int status);
+static void close_cb(uv_handle_t* handle);
+static void shutdown_cb(uv_shutdown_t* req, int status);
+
+static int read_size;
+
+
+static void do_write(uv_tcp_t* handle) {
+  uv_write_t* req;
+  uv_buf_t buf;
+  int i, r;
+
+  buf = uv_buf_init("PING", 4);
+  for (i = 0; i < NUM_WRITE_REQS; i++) {
+    req = malloc(sizeof *req);
+    ASSERT(req != NULL);
+
+    r = uv_write(req, (uv_stream_t*) handle, &buf, 1, write_cb);
+    ASSERT(r == 0);
+  }
+}
+
+
+static void do_close(uv_tcp_t* handle) {
+  if (shutdown_before_close == 1) {
+    ASSERT(0 == uv_shutdown(&shutdown_req, (uv_stream_t*) handle, shutdown_cb));
+    ASSERT(UV_EINVAL == uv_tcp_close_reset(handle, close_cb));
+  } else {
+    ASSERT(0 == uv_tcp_close_reset(handle, close_cb));
+    ASSERT(UV_ENOTCONN == uv_shutdown(&shutdown_req, (uv_stream_t*) handle, shutdown_cb));
+  }
+}
+
+static void alloc_cb(uv_handle_t* handle, size_t size, uv_buf_t* buf) {
+  buf->base = malloc(size);
+  buf->len = size;
+}
+
+static void read_cb2(uv_stream_t* stream, ssize_t nread, const uv_buf_t* buf) {
+  ASSERT((uv_tcp_t*)stream == &tcp_client);
+  if (nread == UV_EOF)
+    uv_close((uv_handle_t*) stream, NULL);
+}
+
+
+static void connect_cb(uv_connect_t* conn_req, int status) {
+  ASSERT(conn_req == &connect_req);
+  uv_read_start((uv_stream_t*) &tcp_client, alloc_cb, read_cb2);
+  do_write(&tcp_client);
+  if (client_close)
+    do_close(&tcp_client);
+}
+
+
+static void write_cb(uv_write_t* req, int status) {
+  /* write callbacks should run before the close callback */
+  ASSERT(close_cb_called == 0);
+  ASSERT(req->handle == (uv_stream_t*)&tcp_client);
+  write_cb_called++;
+  free(req);
+}
+
+
+static void close_cb(uv_handle_t* handle) {
+  if (client_close)
+    ASSERT(handle == (uv_handle_t*) &tcp_client);
+  else
+    ASSERT(handle == (uv_handle_t*) &tcp_accepted);
+
+  close_cb_called++;
+}
+
+static void shutdown_cb(uv_shutdown_t* req, int status) {
+  if (client_close)
+    ASSERT(req->handle == (uv_stream_t*) &tcp_client);
+  else
+    ASSERT(req->handle == (uv_stream_t*) &tcp_accepted);
+
+  shutdown_cb_called++;
+}
+
+
+static void read_cb(uv_stream_t* stream, ssize_t nread, const uv_buf_t* buf) {
+  ASSERT((uv_tcp_t*)stream == &tcp_accepted);
+  read_size += nread;
```

Comment:
This looks wrong. `nread` can be < 0. You handle `UV_EOF` buf if it's any other error, the test is going to act wonky (time out, I suspect.)